app/objects/c_agent.py

```diff
@@ -137,19 +137,41 @@ async def calculate_sleep(self):
         return self.jitter('%d/%d' % (self.sleep_min, self.sleep_max))
 
     async def capabilities(self, ability_set):
+        """Get abilities that the agent is capable of running
+
+        :param ability_set: List of abilities to check agent capability
+        :type ability_set: List[Ability]
+        :return: List of abilities the agents is capable of running
+        :rtype: List[Ability]
+        """
         abilities = []
-        if self.executors:
-            preferred = self._get_preferred_executor()
-            executors = self.executors
-            for ai in set([pa.ability_id for pa in ability_set]):
-                total_ability = [ab for ab in ability_set if (ab.ability_id == ai)
-                                 and (ab.platform == self.platform) and (ab.executor in executors)]
-                if len(total_ability) > 0:
-                    val = next((ta for ta in total_ability if ta.executor == preferred), total_ability[0])
-                    if self.privileged_to_run(val):
-                        abilities.append(val)
+        for ability in ability_set:
+            if self.privileged_to_run(ability) and ability.find_executors(self.executors, self.platform):
+                abilities.append(ability)
         return abilities
 
+    async def get_preferred_executor(self, ability):
+        """Get preferred executor for ability
+
+        Will return None if the agent is not capable of running any
+        executors in the given ability.
+
+        :param ability: Ability to get preferred executor for
+        :type ability: Ability
+        :return: Preferred executor or None
+        :rtype: Union[Executor, None]
+        """
+        preferred_executor_name = self._get_preferred_executor_name()
+        potential_executors = ability.find_executors(self.executors, self.platform)
+
+        if not potential_executors:
+            return
```

Comment:
I would be explicit here and `return None` rather than just a bare `return`. IMO it reads clearer.